common/nvramparser.cpp

```diff
@@ -296,1548 +316,1225 @@ USTATUS NvramParser::parseNvramVolumeBody(const UModelIndex & index)
+            UINT8 calculated = calculateChecksum8(((const UINT8*)storeHeader) + 2, storeHeader->Header.Size - 2);
+            
+            // Add info
+            name = UString("Phoenix EVSA store");
+            info = usprintf("Signature: EVSA\nFull size: %Xh (%u)\nHeader size: %Xh (%u)\nBody size: %Xh (%u)\nAttributes: %08Xh\nReserved: %08Xh\nChecksum: %02Xh",
+                            storeSize, storeSize,
+                            (UINT32)header.size(), (UINT32)header.size(),
+                            (UINT32)body.size(), (UINT32)body.size(),
+                            parsed.attributes(),
+                            parsed.reserved(),
+                            parsed.checksum())
+            + (parsed.checksum() != calculated ? usprintf(", invalid, should be %02Xh", calculated) : UString(", valid"));
+            
+            // Add header tree item
+            UModelIndex headerIndex = model->addItem(localOffset + storeOffset, Types::EvsaStore, 0, name, UString(), info, header, body, UByteArray(), Fixed, index);
+            
+            // Add entries
+            std::map<UINT16, EFI_GUID> guidMap;
+            std::map<UINT16, UString> nameMap;
+            UINT32 entryOffset = parsed.len_evsa_store_header();
+            for (const auto & entry : *parsed.body()->entries()) {
+                UINT8 subtype;
+                UINT32 entrySize;
+                
+                // This is the terminating entry, needs special processing
+                if (entry->_is_null_checksum()) {
+                    // Add free space or padding after all variables, if needed
+                    if (entryOffset < storeSize) {
+                        UByteArray freeSpace = volumeBody.mid(storeOffset + entryOffset, storeSize - entryOffset);
+                        // Add info
+                        info = usprintf("Full size: %Xh (%u)", (UINT32)freeSpace.size(), (UINT32)freeSpace.size());
+                        
+                        // Check that remaining unparsed bytes are actually empty
+                        if (freeSpace.count(emptyByte) == freeSpace.size()) { // Free space
+                            // Add tree item
+                            model->addItem(entryOffset, Types::FreeSpace, 0, UString("Free space"), UString(), info, UByteArray(), freeSpace, UByteArray(), Fixed, headerIndex);
+                        }
+                        else {
+                            // Add tree item
+                            model->addItem(entryOffset, Types::Padding, getPaddingType(freeSpace), UString("Padding"), UString(), info, UByteArray(), freeSpace, UByteArray(), Fixed, headerIndex);
+                        }
+                    }
                     break;
                 }
-            if (!reservedBytesValid)
-                continue;
+                
+                const EVSA_ENTRY_HEADER* entryHeader = (const EVSA_ENTRY_HEADER*)(volumeBody.constData() + storeOffset + entryOffset);
```

Comment:
I might be missing the model here, but why is it safe to expect an object of EVSA_ENTRY_HEADER size at entryOffset?

Reply:
This is guaranteed by KaitaiStream source